### PR TITLE
Add cosine_with_min_lr_schedule_with_warmup_lr_rate scheduler in Trainer

### PR DESCRIPTION
**Pull request by richardodliu**
# What does this PR do?

Add cosine_with_min_lr_schedule_with_warmup_lr_rate scheduler in Trainer, which is based on https://github.com/huggingface/transformers/pull/29341.

As mentioned in the previous PR, it implemented "a warmup period during which it increases linearly between 0 and the initial learning rate set in the optimizer." However, I recently investigated the DeepSpeed framework and noticed that their function implements additional features, namely "warmup_min_ratio" in https://github.com/richardodliu/DeepSpeed/blob/master/deepspeed/runtime/lr_schedules.py#L774, which supports a warmup start learning rate ratio differs from 0. Considering that DeepSpeed is a crucial component framework in the implementation of Transformers, I aim to ensure consistency between the two in the implementation of learning rate schedulers through this PR. This is to prevent potential confusion for users who employ both frameworks simultaneously.

Our implementation is based on improvements from previous PR, providing significant benefits. Specifically, we allow the reuse of this method without modifying any input parameters(Don't worry; if you prefer to specify the parameter, you are certainly allowed to do so by simply passing it as an argument "warmup_lr_rate", which is used to specify the ratio between the start learning rate and the initial learning rate). In such cases, the implementation is equivalent to setting "warmup_lr_rate" to "1/warmup_steps". Regarding enhancements to the training process, since it is recommended to perform "optimizer.step()" before "lr_scheduler.step()", the learning rate for the batch corresponding to the first step would be zero under the previous implementation. This means that the gradients would not be updated for that batch, effectively wasting it. Our updated method addresses this issue. While this phenomenon may not be noticeable with larger dataset sizes, it becomes significant with smaller datasets where the total number of steps is limited. Additionally, our implementation ensures that the final small learning rate is reachable, rather than being updated only after the training is completed. Overall, our approach is better suited as an improvement to the existing method rather than a complete overhaul. Therefore, we implemented our idea by creating a new function, allowing users the flexibility to choose which implementation method they prefer.

Fixes: WarmupCosineLR missed in WarmupCosineLR

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@muellerzr and @SunMarc

### CODE REVIEW

**Comment by richardodliu** [comment]
@muellerzr and @SunMarc : Hello, I submitted this PR over a week ago, but it hasn't been reviewed yet. I've checked the unit test results and confirmed that the failing tests don't seem to be caused by my code changes. Additionally, the test cases I've submitted are passing. Therefore, I believe this PR is ready for review.

**Review by SunMarc — APPROVED** [review]
LGTM ! Thanks for adding this @richardodliu !

**Comment by richardodliu** [comment]
@muellerzr and @ArthurZucker: Hello, I submitted this PR over three weeks ago, and @SunMarc has reviewed the PR. I've checked the unit test results and confirmed that the failing tests don't seem to be caused by my code changes. Additionally, the test cases I've submitted are passing. Therefore, I believe this PR is ready for merge.

**Review by muellerzr — APPROVED** [review]
Thanks, sorry for the delay. @amyeroberts is our final judgement/on watch :)

Thanks for the addition

File: src/transformers/optimization.py [hunk 1]
```diff
@@ -387,6 +387,84 @@ def get_cosine_with_min_lr_schedule_with_warmup(
...
+    *,
+    num_warmup_steps: int,
+    num_training_steps: int,
+    num_cycles: float,
+    min_lr_rate: float = 0.0,
+    warmup_lr_rate: float = None,
+):
+    if current_step < num_warmup_steps:
+        if warmup_lr_rate is None:
+            return float(current_step + 1) / float(max(1, num_warmup_steps))
```
Comment (amyeroberts) [inline src/transformers/optimization.py:401]: Couldn't we place some input validation on `num_warmup_steps` in `get_cosine_with_min_lr_schedule_with_warmup_lr_rate` s.t. we know it's and int >= 1?

File: src/transformers/optimization.py [hunk 2]
```diff
@@ -387,6 +387,84 @@ def get_cosine_with_min_lr_schedule_with_warmup(
...
+        else:
+            return float(warmup_lr_rate) + float(1.0 - warmup_lr_rate) * float(current_step) / float(
+                max(1, num_warmup_steps - 1)
+            )
+    progress = float(current_step - num_warmup_steps + 1) / float(max(1, num_training_steps - num_warmup_steps))
+    factor = 0.5 * (1.0 + math.cos(math.pi * float(num_cycles) * 2.0 * progress))
```
Comment (amyeroberts) [inline src/transformers/optimization.py:407]: num_cycles should already be a float according to the type hints

File: src/transformers/optimization.py [hunk 3]
```diff
@@ -387,6 +387,84 @@ def get_cosine_with_min_lr_schedule_with_warmup(
...
+    factor = factor * (1 - min_lr_rate) + min_lr_rate
+    return max(0, factor)
```
Comment (amyeroberts) [inline src/transformers/optimization.py:409]: Rather than repeated `float` calls - this would be easier to read if we just cast at the start of the function


```suggestion
    current_step = float(current_step)
    num_warmup_steps = float(num_warmup_steps)
    num_training_steps = float(num_training_steps)
    
    if current_step < num_warmup_steps:
        if warmup_lr_rate is None:
            return (current_step + 1.) / max(1., num_warmup_steps)
        else:
            warmup_lr_rate = float(warmup_lr_rate)
            return warmup_lr_rate + (1.0 - warmup_lr_rate) * (current_step) / (max(1, num_warmup_steps - 1)
    progress = (current_step - num_warmup_steps + 1.) / (max(1., num_training_steps - num_warmup_steps))
    factor = 0.5 * (1.0 + math.cos(math.pi * num_cycles * 2.0 * progress))
    factor = factor * (1 - min_lr_rate) + min_lr_rate
    return max(0, factor)
```

File: src/transformers/optimization.py [hunk 4]
```diff
@@ -387,6 +387,84 @@ def get_cosine_with_min_lr_schedule_with_warmup(
...
+    warmup_lr_rate: float = None,
+):
+    if current_step < num_warmup_steps:
+        if warmup_lr_rate is None:
+            return float(current_step + 1) / float(max(1, num_warmup_steps))
+        else:
+            return float(warmup_lr_rate) + float(1.0 - warmup_lr_rate) * float(current_step) / float(
+                max(1, num_warmup_steps - 1)
+            )
+    progress = float(current_step - num_warmup_steps + 1) / float(max(1, num_training_steps - num_warmup_steps))
```
Comment (amyeroberts) [inline src/transformers/optimization.py:406]: Similarly here - we could add input validation to ensure that num_training_steps > num_warmup_steps instead of needing this `max` fallback

**Review by amyeroberts — APPROVED** [review]
Thanks for adding this! 

Just some nits

**Comment by richardodliu** [comment]
I'm sorry I didn't check the relevant comments in time last fortnight due to my personal commitments, is there anything else I should be doing now? @ArthurZucker

**Comment by muellerzr** [comment]
@richardodliu do `pip install -e .[quality]; make fixup` and this should make the code quality tests pass!

**Comment by ArthurZucker** [comment]
@bot /style

**Comment by ArthurZucker** [comment]
@bot /style

File: src/transformers/optimization.py [hunk 5]
```diff
@@ -384,6 +384,87 @@ def get_cosine_with_min_lr_schedule_with_warmup(
...
+
+def get_cosine_with_min_lr_schedule_with_warmup_lr_rate(
+    optimizer: Optimizer,
+    num_warmup_steps: int,
+    num_training_steps: int,
+    num_cycles: float = 0.5,
+    last_epoch: int = -1,
+    min_lr: float = None,
+    min_lr_rate: float = None,
+    warmup_lr_rate: float = None,
```
Comment (ArthurZucker) [inline src/transformers/optimization.py]:
```suggestion
    min_lr: Optional[float] = None,
    min_lr_rate: Optional[float] = None,
    warmup_lr_rate: Optional[float] = None,
```

File: src/transformers/optimization.py [hunk 6]
```diff
@@ -384,6 +384,87 @@ def get_cosine_with_min_lr_schedule_with_warmup(
...
 
 
+def _get_cosine_with_min_lr_schedule_with_warmup_lr_rate_lambda(
+    current_step: int,
+    *,
+    num_warmup_steps: int,
+    num_training_steps: int,
+    num_cycles: float,
+    min_lr_rate: float = 0.0,
+    warmup_lr_rate: float = None,
```
Comment (ArthurZucker) [inline src/transformers/optimization.py]:
```suggestion
    warmup_lr_rate: Optional[float] = None,
```

**Comment by ArthurZucker** [comment]
@bot /style
